File: research/lev_diag_performance/test_levdiag.rb



$MAX_VALUE = 1000000000
 
# delta is number of row elements to be calculated above and below the central  
# diagonal element  

def calculate_diag(text, pattern, delta)
    # defines size of matrix 
    rows, cols = text.length, pattern.length    
    
    # initialize column data 
    colsize = delta * 2 + 1     # number of row elements to be calculated per column
   
    # init column arrays to keep prev. calculated results and current results
    lcol = Array.new(colsize + 3, $MAX_VALUE)
    rcol = Array.new(colsize + 3, $MAX_VALUE)
    
    # init left column data
    for i in 0..(delta + 2) 
        lcol[i] = i
    end

    # initialize array to keep result data
    # this is the plain array where every calculated row diagonal 
    # elements are placed in fixed sized slot - (colsize + 1). The first (extra)
    # element in the slot is used to keep first row index the data have been
    # calculated. array is pre-filled with MAX_VALUE to make difference between
    # elements that have been calculated and untouchable elements.  
    data = Array.new((cols * (colsize + 1)), $MAX_VALUE)
    
    # initialize start and end rows indexes
    l_row1, l_row2, l_col = 0, delta + 1, -1 
    
    # slightly speed up access to variable
    maxrow = rows - 1

    for col in 0..(cols-1)
        # calculate anchor row index
        arow = (col * rows) / cols
        
        # upper row
        r_row1 = arow > delta  ? arow - delta : 0
        
        # lower row
        r_row2 = arow + delta
        if (r_row2 >= rows) 
            r_row2 = maxrow 
        end
        
        lcol[0] = l_row1 == 0 ? col     : $MAX_VALUE   
        rcol[0] = r_row1 == 0 ? col + 1 : $MAX_VALUE
        
        # to speed up access to array element store it in temporary var
        col_ch = pattern[col]
        
        # pre-calculated variables slightly speed up loop below
        drr, dlr = 1 - r_row1, 1 - l_row1    
        
        # calculate starting index for the given column 
        index = col * (colsize + 1)
        
        # store starting row to be calculated as the first element 
        data[index] = r_row1
        index += 1
        
        # calculate diagonal elements for the given column
        for row in r_row1..r_row2 
            # again pre-compute values to speed up calculation
            i, j = row + drr, row + dlr    
                        
            # compare text and pattern characters 
            if (text[row] == col_ch) 
                rcol[i] = lcol[j - 1]
                data[index] = rcol[i]
            else  
                # the construction below is equivalent of min (a,b,c) 
                # function just to win another thousand milliseconds
                a,b,c = lcol[j], rcol[i - 1], lcol[j - 1]
                
               # puts a, b, j
                
                if (a < b)
                    if (c < a) 
                        a = c
                    end     
                else
                    a = b < c  ? b : c
                end
           
                a += 1              
                rcol[i] = data[index] = a 
            end

            index += 1    
        end
        
        # replace left row reference with just calculated right row
        # since for the next column calculation current right row should 
        # be considered as left row.
        a, lcol = lcol, rcol
        rcol = a
        l_row1, l_row2 = r_row1, r_row2
    end
    
    return data
end


#
# dirty code just to visualize results
#
def print_result(s1, s2, delta, res)
    rows, cols = s1.length, s2.length
    colsize = delta * 2 + 1
 
    buf = ''
    for row in 0..(rows-1)
        puts row
        for col in 0..(cols-1)
            index  = (colsize + 1) * col
            srow   = res[index]
            index  += 1
            i      = index + row - srow
            
            if (row < srow || row >= srow + colsize || res[i] == $MAX_VALUE) 
                buf += " xx "
            else
                buf += " " + res[i].to_s + " "      
            end
         end
     end       
        buf += "\n"
                
    print buf
end

def test1()
    # test input data
    delta = 3
    s1    = "asdffghjjkll"
    s2    = "sxcvbbnhmkjkjkjhjkg"

    r = calc(s1, s2, delta)
    print_result(s1, s2, delta, r)
end

def test2()
    # load input data located in file  
    delta = 100 
    s1 = IO.readlines('test.txt','r:UTF-8').to_s 
    s2 = IO.readlines('test2.txt','r:UTF-8').to_s 
    calc(s1, s2, delta)
end

def calc(s1, s2, delta)
    puts "Input data : s1.len = " + s1.length.to_s + ", s2.len = " + s2.length.to_s + ", delta = " + delta.to_s

    # calculate diagonal
    t = Time.new
    r = calculate_diag(s1, s2, delta)
    puts (Time.new - t)

    # last element index
    i = (delta * 2 + 2) * (s2.length-1) + delta + 1
    puts "Last elements: ", r[i], r[i-1] 
        
    return r
end




test2()

File: research/lev_diag_performance/test_levdiag.cpp
#include <string>
#include <vector>
#include <iostream>
#include <fstream>

using namespace std;

#define MAX_VALUE 10000000

typedef vector< int > INTVECT;

//
// Calculate diag
//
INTVECT calculate_diag(const string& text, const string& pattern, const int& delta) {
  
  int rows    = text.length(), cols = pattern.length();
  int colsize = delta * 2 + 1;

  // init two columns for calculation and fill it with max value   
  INTVECT lcol;
  INTVECT rcol;
  lcol.resize( colsize + 3, MAX_VALUE );
  rcol.resize( colsize + 3, MAX_VALUE );

  // setup initial values for left column
  for (int row=0; row < delta + 3; row++) {
      lcol[row] = row;
  }
  
  INTVECT data;
  data.resize( cols * (colsize + 1), MAX_VALUE );
      
  // initialize start and end rows indexes
  int l_row1 = 0, l_row2 = delta + 1, l_col = -1; 
          
  // slightly speed up access to variable
  int maxrow = rows - 1;
      
  for (int col=0; col < cols; col++) {
    // calculate anchor row index
    int arow = (col * rows) / cols;
    
    // upper row
    int r_row1 = arow > delta ? arow - delta : 0;
    
    // lower row
    int r_row2 = arow + delta;
    if (r_row2 >= rows) r_row2 = maxrow; 

    lcol[0] = l_row1 == 0 ? col     : MAX_VALUE;   
    rcol[0] = r_row1 == 0 ? col + 1 : MAX_VALUE;
    
    // to speed up access to array element store it in temporary var
    char col_ch = pattern[col];
          
    // pre-calculated variables slightly speed up loop below
    int drr = 1 - r_row1;
    int dlr = 1 - l_row1;    
    
    // calculate starting index for the given column 
    int index = col * (colsize + 1);
    
    // store starting row to be calculated as the first element 
    data[index] = r_row1;
    index ++;
                
    // calculate diagonal elements for the given column
    for (int row = r_row1; row < r_row2 + 1; row++) { 
      // again pre-compute values to speed up calculation
      int i = row + drr; 
      int j = row + dlr;    
                  
      // compare text and pattern characters 
      if (text[row] == col_ch) { 
        rcol[i] = lcol[j - 1];
        data[index] = rcol[i];
      }    
      else {  
        // the construction below is equivalent of min (a,b,c) 
        // function just to win another thousand milliseconds
        int a = lcol[j], b = rcol[i - 1], c = lcol[j - 1];
        if (a < b) {
            if (c < a) a = c; 
        }
        else {
            a = b < c ? b : c;
        }

        a++;              
        rcol[i] = data[index] = a; 
      }
           
      index ++;    
    }

    // replace left row reference with just calculated right row
    // since for the next column calculation current right row should 
    // be considered as left row.
    INTVECT a = lcol;
    lcol = rcol;
    rcol = a;
    l_row1 = r_row1; 
    l_row2 = r_row2;
  }
  
  return data;
}
  
//
// dirty code just to visualize results
//
void print_result(const string& s1, const string& s2, const int& delta, const INTVECT& res) {
  int rows = s1.length(), cols = s2.length();
  int colsize = delta * 2 + 1;

  for (int row = 0; row < rows; row++) {
    for (int col=0; col < cols; col++) {
      int index  = (colsize + 1) * col;
      int srow   = res[index];
      index ++;
      int i = index + row - srow;
      
      if (row < srow || row >= srow + colsize || res[i] == MAX_VALUE) { 
        cout << " xx ";
      }    
      else {
        cout << " " << res[i] << " ";
      }        
    }
    cout << endl; 
  }                
  cout << endl;
}

string readFile(const string& filename) {
  string allstr = "";
  string str;
  ifstream infile;
  infile.open( filename.c_str() );
  
  if ( !infile.is_open() ) {
    cout << "file doesn't exist" << endl;
    exit( -1 );
  }
  else {
    infile >> allstr; 
    while( !infile.eof() ) {
      infile >> str; 
      allstr += (" " + str);
    }
    infile.close();
    return allstr;    
  }
}


int main( int argc, char* argv[] ) {
  
//  string s1  = "sdasdasd";
//  string s2 = "sdsadasd"; 
//  int    d  = 2;
  
  clock_t t1, t2;  
  
  string s1 = readFile("test.txt");
  string s2 = readFile("test2.txt"); 
  int     d = 100;

  cout << "s1.length = " << s1.length() << "  s2.length = " << s2.length() << endl;

  t1 = clock();
  
  INTVECT r = calculate_diag(s1, s2, d);
//  print_result(s1, s2, d, r);  

  t2 = clock();
  float diff = ((float)t2 - (float)t1) / 1000000.0F;
  
  cout << "runtime: " << diff << " ms" << endl;

  int li = (d * 2 + 2) * (s2.length() - 1 ) + d;
  cout << "Last element: " << r[li] << endl;
  
}

